### ISSUE TITLE
release build - R8 reports Missing class com.google.firebase.messaging.TopicOperation$TopicOperations

### ISSUE THREAD

**Issue by mikehardy** [firebase/firebase-android-sdk#3740]
### [READ] Step 1: Are you in the right place?

Issues filed here should be about bugs in __the code in this repository__. 
If you have a general question, need help debugging, or fall into some
other category use one of these other channels:

  * For general technical questions, post a question on [StackOverflow](http://stackoverflow.com/)
    with the firebase tag.
  * For general Firebase discussion, use the [firebase-talk](https://groups.google.com/forum/#!forum/firebase-talk)
    google group.
  * For help troubleshooting your application that does not fall under one
    of the above categories, reach out to the personalized
    [Firebase support channel](https://firebase.google.com/support/).

### [REQUIRED] Step 2: Describe your environment

  * Android Studio version: gradle plugin 7.0.4 (that's the thing that contains R8)
  * Firebase Component: Messaging
  * Component version: BoM 30.0.1

#### Steps to reproduce:

Do a release build that includes messaging, on android, with proguard / code shrinking enabled

#### Relevant Code:

Again, with a code shrink issue I don't *think* code is required, I *think* that you only need to include the library (like in a quickstart) and then build for release

```

> Task :app:minifyProdReleaseWithR8
WARNING:Missing classes detected while running R8. Please add the missing classes or apply additional keep rules that are generated in /Users/mike/work/Kullki/ksocialscore/packages/public-app/android/app/build/outputs/mapping/prodRelease/missing_rules.txt.

WARNING:R8: Missing class com.google.firebase.messaging.TopicOperation$TopicOperations (referenced from: void com.google.firebase.messaging.TopicOperation.<init>(java.lang.String, java.lang.String))

```

I don't see a proguard-rules.pro file in the source tree here so I can't say for certain if it is incorrect rules, or just that there should be a proguard file and there isn't one at all

**Comment by sgjesse** [comment]
Looking at https://mvnrepository.com/artifact/com.google.firebase/firebase-messaging/23.0.5, the code for the private `com.google.firebase.messaging.TopicOperation(String, String)` constructor looks like this:
```
  private com.google.firebase.messaging.TopicOperation(java.lang.String, java.lang.String);
    descriptor: (Ljava/lang/String;Ljava/lang/String;)V
    flags: (0x0002) ACC_PRIVATE
    Code:
      stack=3, locals=3, args_size=3
         0: aload_0
         1: invokespecial #27                 // Method java/lang/Object."<init>":()V
         4: aload_0
         5: aload_2
         6: aload_1
         7: invokestatic  #31                 // Method normalizeTopicOrThrow:(Ljava/lang/String;Ljava/lang/String
;)Ljava/lang/String;
        10: putfield      #33                 // Field topic:Ljava/lang/String;
        13: aload_0
        14: aload_1
        15: putfield      #35                 // Field operation:Ljava/lang/String;
        18: aload_0
        19: new           #37                 // class java/lang/StringBuilder
        22: dup
        23: invokespecial #38                 // Method java/lang/StringBuilder."<init>":()V
        26: aload_1
        27: invokevirtual #42                 // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/l
ang/StringBuilder;
        30: ldc           #9                  // String !
        32: invokevirtual #42                 // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/l
ang/StringBuilder;
        35: aload_2
        36: invokevirtual #42                 // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/l
ang/StringBuilder;
        39: invokevirtual #46                 // Method java/lang/StringBuilder.toString:()Ljava/lang/String;
        42: putfield      #48                 // Field serializedString:Ljava/lang/String;
        45: return
      LineNumberTable:
        line 43: 0
        line 44: 4
        line 45: 13
        line 46: 18
        line 47: 45
      LocalVariableTable:
        Start  Length  Slot  Name   Signature
            0      46     0  this   Lcom/google/firebase/messaging/TopicOperation;
            0      46     1 operation   Ljava/lang/String;
            0      46     2 topic   Ljava/lang/String;
    RuntimeInvisibleParameterAnnotations:
      parameter 0:
        0: #24()
          com.google.firebase.messaging.TopicOperation$TopicOperations
      parameter 1:
```
There is a reference to the class `com.google.firebase.messaging.TopicOperation$TopicOperations` as a parameter annotation, but that class is not present in the `classes.jar` from the `aar`.

Adding the rule `-dontwarn com.google.firebase.messaging.TopicOperation$TopicOperations` should be fine on the library consumer side.

On the library building side it should be checked if the annotation is needed and if not ensure that is is removed or otherwise include the annotation class.

I used `javap -v -p -cp classes.jar com.google.firebase.messaging.TopicOperation` to get the byte code above.

**Comment by argzdev** [comment]
Thanks for reporting, @mikehardy. I created a clean project with the configuration you've provided (also tried with a quickstart project), however I'm unable to reproduce the issue.

Setup:
- Android Gradle Plugin Version: `7.0.4` (I also tried `7.2.0`)
- Gradle Version: `7.3.3`

Relevant code:
```
{
    ...
    buildTypes {
        release {
            minifyEnabled true
            proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
        }
    }
    ...
}

dependencies {
    ...
    implementation platform('com.google.firebase:firebase-bom:30.0.1')
    implementation 'com.google.firebase:firebase-messaging-ktx'
}
```

Run command in gradle: `gradle app:minifyReleaseWithR8`

Am I missing anything? Could you provide more details, or a minimal repro of the issue? It'll greatly help us investigate this further. Thanks!

**Comment by mikehardy** [comment]
Here's a fully-scripted demo, I'm mostly in the react-native (i.e., react-native-firebase) ecosystem so my easiest repro tools are mostly based on that and I reach for them first, but I did my best to strip it down to android only (or, mostly), though it will require node v16 to run, and expects things to be unix-y (linux or macOS should work, WSL can work but it's fiddly)

https://github.com/mikehardy/rnfbdemo/blob/006d6a4d3c7b88fd5d693b345a20304e44b89ef8/make-demo.sh

It shows this for me
> WARNING:R8: Missing class com.google.firebase.messaging.TopicOperation$TopicOperations (referenced from: void com.google.firebase.messaging.TopicOperation.<init>(java.lang.String, java.lang.String))

If for some reason this repro doesn't work I can try to go off a quickstart or something, but hopefully it works :crossed_fingers:  for repro. Thanks in advance for the attention

**Comment by argzdev** [comment]
Thanks for the repro, @mikehardy. Unfortunately, after generating the repo, I'm having problems running it natively, specifically the `gradle app:minifyReleaseWithR8` in `rnfbdemo/android` in Android studio. I changed the `def enableProguardInReleaseBuilds = true` in the gradle for R8 support, but I'm getting issues `Execution failed for task ':app:bundleReleaseJsAndAssets'.`. I'm not exactly sure with the issue here, but could you instead provide another repro using native code? Or the [quickstart](https://github.com/firebase/quickstart-android) would also work nicely. Thanks in advance!

**Comment by mikehardy** [comment]
Please note the command I crafted carefully for you ;-)

> ./gradlew clean -x :app:bundleReleaseJsAndAssets :app:minifyReleaseWithR8 --rerun-tasks

**Comment by argzdev** [comment]
Sorry! I've overlooked this from the script file you've provided. I've tried running it, but it worked.

```
BUILD SUCCESSFUL in 1m 8s
320 actionable tasks: 320 executed

Build Analyzer results available
2:03:06 AM: Tasks execution finished 'clean :app:minifyReleaseWithR8 --rerun-tasks --exclude-task :app:bundleReleaseJsAndAssets'.
```

Could you try using a higher version of gradle version or AGP version?

This is the current version I'm using:
- Android Gradle Plugin Version: `7.1.1`
- Gradle Version: `7.3.3`

**Comment by mikehardy** [comment]
Fantastic! Thank you :-) - I was actually responding to original version of comment when github helpfully async-updated the page as I was typing. Cheers

**Comment by richanshah** [comment]
does any one found solution?

**Comment by mikehardy** [comment]
@richanshah did you check the linked PR? It's been fixed already. Did you look at the specific commit for the PR? https://github.com/firebase/firebase-android-sdk/pull/3746/commits/b79c8fa8dd04e0ff9e0f020b063fbaa007451cb1 - it just hasn't been released yet. So the solution is to wait. Additionally, it was a warning and it was actually just an annotation that wasn't necessary in the built product, so other than generating some build noise it is inconsequential and may be ignored if I understand correctly

**Comment by richanshah** [comment]
I'll check thanks

Richa Shah
Android Developer
phone:  9016804343
email:  ***@***.***

**Comment by gsakakihara** [comment]
This should be fixed with version 23.0.6.
